core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -2666,4 +2666,13 @@ package object config {
       .version("4.0.0")
       .booleanConf
       .createWithDefault(false)
+
+  private[spark] val SPARK_SHUTDOWN_TIMEOUT_MS =
+    ConfigBuilder("spark.shutdown.timeout")
+      .doc("Defines the timeout period to wait for all shutdown hooks to be executed. " +
+        "This must be passed as a system property argument in the Java options, for example " +
+        "spark.driver.extraJavaOptions=\"-Dspark.shutdown.timeout=60s\".")
```

Comment:
According to the PR description's claim, you cannot achieve this with `spark.driver.extraJavaOptions`, @robreeves ?

> This is not easily configurable before this change. The underlying org.apache.hadoop.util.ShutdownHookManager has a default timeout of 30 seconds. It can be configured by setting hadoop.service.shutdown.timeout, but this must be done in the core-site.xml/core-default.xml because a new hadoop conf object is created and there is no opportunity to modify it.

Reply:
Sorry, but I'm not sure what you mean. I will try to clarify. Before this change there is no easy way to set the timeout easily besides in the `core-site.xml` file using `hadoop.service.shutdown.timeout`. For my use case this is problematic because this is centrally managed and not Spark specific. 

During initialization Spark will register all `spark.*` properties as system properties. Then SparkConf will read them each time an object is created. The problem here is `ShutdownHookManager` is called so early during start up the Spark properties have not been registered as system properties. The workaround is for the configuration to explicitly pass the config as a system property to Java using `-D` like in the example.